feat(dynamodb): allow registration of custom type marshal/unmarshal functions

ztison opened the pull request:
I have implemented a basic functionality for registering custom-type marshalers. The code expects the exact type match. Users can always register the same marshaler for related types. It would be more efficient than iterating over the map and checking if a type is assignable to the interface.

Commit:
Needs debug for s3100continue middleware

jfcg committed:
Fix staticcheck linter warnings (#2044)

Commit:
Update API model

Commit:
Update endpoints model

Commit:
Regenerated Clients

Commit:
Release 2023-03-10

Commit:
Port aws-sdk-go-v1's {Expect: 100-continue} http request header customization middleware for s3 client http PUT operation larger than 2MB or with unknown size Body.

Commit:
changelog to s3 100continue customzaiton commit

Commit:
codegen updated module dependencies versions

Commit:
Modify some comments in s3 100-continue middleware

Commit:
Update endpoints model

Commit:
Update API model

Commit:
Regenerated Clients

Commit:
Release 2023-03-13

aajtodd committed:
breakfix: sparse trait missing on support service (#2050)

Sync models and regenerate to pickup missing sparse trait.
This is a breakfix change to properly support the following operations:

* `DescribeTrustedAdvisorCheckRefreshStatuses`
* `DescribeTrustedAdvisorCheckSummaries`

ticket: P83022241

aajtodd committed:
feat: add flag to disable IMDSv1 fallback (#2048)

Commit:
Update endpoints model

Commit:
Update API model

Commit:
Regenerated Clients

Commit:
Release 2023-03-14

Commit:
Update API model

Commit:
Update endpoints model

Commit:
Regenerated Clients

Commit:
Release 2023-03-15

Commit:
Modify some logic in s3 100-continue middleware PR

Commit:
Modify some logic in s3 100-continue middleware PR

Commit:
Merge remote-tracking branch 'origin/main' into add100ContinueCustomization

# Conflicts:
#	config/go.mod

Commit:
Add comment to s3100continue

aajtodd committed:
delay WebIdentityRoleOptions validation (#2030)

When loading the default config and `AWS_WEB_IDENTITY_TOKEN_FILE` is set it
triggers the assume role with web identity provider. This provider
requires it's configuration either entirely in environment variables or
shared config. The Go SDK offers an API to configure this provider
further via code which was being ignored or causing early validation
issues. This change allows `AWS_WEB_IDENTITY_TOKEN_FILE` to be set
and all other options to be supplied by Go code relaxing the constraint
that everything has to come from the environment.

fixes aws/aws-sdk-go-v2#2015

Commit:
Update API model

Commit:
Update endpoints model

Commit:
Regenerated Clients

Commit:
Release 2023-03-16

Commit:
Merge branch 'main' into add100ContinueCustomization

Commit:
Update API model

Commit:
Update endpoints model

Commit:
Regenerated Clients

Commit:
Release 2023-03-17

Commit:
Merge branch 'main' into add100ContinueCustomization

Commit:
Enable user configuration of Expect: s3100continue header customization, exempt Expect header from signV4

Commit:
Add changlog to git push baf538a225

ztison committed:
Custom type marshal/unmarshal functions

Commit:
Update API model

Commit:
Update endpoints model

Commit:
Regenerated Clients

Commit:
Release 2023-03-20

austince commented on feature/dynamodb/attributevalue/decode.go:
```diff
@@ -896,6 +946,22 @@ func indirect(v reflect.Value, opts indirectOptions) (Unmarshaler, reflect.Value
+	defer d.unmarshalersLock.RUnlock()
+	if len(d.unmarshalers) == 0 {
+		return nil
+	}
```
nit, I think this could be skipped as reading from a `nil` map is ok.

ztison replied:
Skipped

austince commented on feature/dynamodb/attributevalue/decode_test.go:
```diff
@@ -2,14 +2,13 @@ package attributevalue
 	"fmt"
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/dynamodb/types"
+	"github.com/google/go-cmp/cmp"
```
nit, probably move back

austince commented on feature/dynamodb/attributevalue/encode.go:
```diff
@@ -433,15 +464,12 @@ func (e *Encoder) encode(v reflect.Value, fieldTag tag) (types.AttributeValue, e
-			return nil, err
-		} else if av != nil {
-			return av, nil
+		if av, err := e.tryMarshaler(v); err != nil || av != nil {
+			return av, err
```
nit, maybe leave as was to reduce diff? else ifs seem to be their style

austince commented on feature/dynamodb/attributevalue/encode_test.go:
```diff
@@ -2,14 +2,13 @@ package attributevalue
 import (
 	"fmt"
+	smithydocument "github.com/aws/smithy-go/document"
+	"github.com/google/go-cmp/cmp/cmpopts"
```
nit to move back

Commit:
Merge remote-tracking branch 'origin/main' into add100ContinueCustomization

austince reviewed (approved):
Looks great, nice work. You might also look into writing a benchmark for the encoder/decoder? https://github.com/aws/aws-sdk-go-v2/blob/9af2842ca9258238ee5f72d85e5d61b12c4a7737/feature/dynamodb/attributevalue/marshaler_test.go#L522-L547

Commit:
Change s3 100continue middleware codegen comment, shrink middleware scale to PutObject and UploadPart

Commit:
add changelog to last commit

Commit:
Change some variable name and use operation shape id to represent operation shape

Commit:
Merge remote-tracking branch 'origin/add100ContinueCustomization' into add100ContinueCustomization

Commit:
Keep one changelog for PR

Commit:
Keep one changelog for PR

wty-Bryant committed:
Merge pull request #2051 from aws/add100ContinueCustomization

Port v1 SDK customization for s3 HTTP PUT request

ztison commented:
@austince THanks for the review. Good idea to add a bench. I added it.

Commit:
Update API model

Commit:
Update endpoints model

Commit:
Regenerated Clients

Commit:
Release 2023-03-21

Commit:
Update API model

Commit:
Update endpoints model

Commit:
Regenerated Clients

Commit:
Release 2023-03-22

Commit:
Update API model

Commit:
Update endpoints model

Commit:
Regenerated Clients

Commit:
Release 2023-03-23

Commit:
Add ApiGatewayExportsNullabilityExceptionIntegration (#2056)

Due to internal model fixes for API Gateway (APIGW) exports, certain shapes
and members that target those shapes that used to have defaults became
nullable. In order to not break the existing Go v2 client API nullability,
this customization adds back default values for affected shapes:

- Root Boolean shapes: false
- Root Number shapes: 0
- Snapshotted Members shapes: inherit defaults

The class of services affected are APIGW services and APIGW services that
migrated to Smithy, seen in `APIGW_NULLABILITY_EXCEPTION_SERVICES`.

A "snapshot" of root-level and member shapes are captured in
`APIGW_exports_nullability_exceptions.json`, a mapping of service shape IDs
to a list of affected root-level and member shapes.

Cases:
- Snapshotted Root level:
  - If a snapshotted root level shape doesn't exist, an exception will be
    thrown.
  - Otherwise, a default will be patched if a default doesn't already exist
- Snapshotted Member level:
  - If a snapshotted member level shape doesn't exist, an exception will be
    thrown.
  - Otherwise, a default will be patched if a default doesn't already exist
- Nonsnapshotted Member level:
  - All nonsnapshotted member level shape that target a snapshotted root level
    will be identified and throw an error.
  - This is prevent breaking changes if a nonsnapshotted member changes target
    from a snapshotted root shape to a nonsnapshotted root shape, as the SDK
    has no prior info on either shape.

Also removes `serviceid.smithy` which was not used anywhere.

Commit:
Update endpoints model

Commit:
Update API model

Commit:
Regenerated Clients

Commit:
Release 2023-03-24

Commit:
Update API model

Commit:
Regenerated Clients

Commit:
Release 2023-03-27

Commit:
Update API model

Commit:
Update endpoints model

Commit:
Regenerated Clients

Commit:
Release 2023-03-28

Commit:
Update API model

Commit:
Update endpoints model

Commit:
Regenerated Clients

Commit:
Release 2023-03-29

Commit:
Update API model

Commit:
Update endpoints model

Commit:
Regenerated Clients

Commit:
Release 2023-03-30

Commit:
Update endpoints model

Commit:
Update API model

Commit:
Regenerated Clients

Commit:
Release 2023-03-31

Commit:
Update API model

Commit:
Update endpoints model

Commit:
Regenerated Clients

Commit:
Release 2023-04-03

Commit:
Update API model

Commit:
Update endpoints model

Commit:
Regenerated Clients

Commit:
Release 2023-04-04

Commit:
Update API model

Commit:
Update endpoints model

Commit:
Regenerated Clients

Commit:
Release 2023-04-05

Commit:
Update API model

Commit:
Update endpoints model

Commit:
Regenerated Clients

Commit:
Release 2023-04-06

sbiscigl committed:
Add announcement for next release for dlm

rem7 committed:
NXDOMAIN errors should not be retried (#2083)

The standard retryer keeps retrying domains that do not resolve. This
hangs the execution until retries are exhausted.
NXDOMAIN or 'no such host' errors are not transient and should fail immediately.
An example would be when a user enters an invalid region that doesn't resolve.

Commit:
fix APIGW exports nullability exceptions

Commit:
Update API model

Commit:
Regenerated Clients

Commit:
Release 2023-04-07

Commit:
Update aws-sdk-go-v2's comment codegened from Smithy Go's updated document smart wrapping feature

sbiscigl committed:
fix APIGW exports nullability exceptions

wty-Bryant committed:
Merge pull request #2089 from aws/auditAccessibility

Update aws-sdk-go-v2's comment codegened from Smithy Go's updated doc…

Commit:
Skip tests for ApiGatewayExportsNullabilityExceptionIntegration

Commit:
Add logging for tests

Commit:
Update endpoints model

Commit:
Update API model

Commit:
Regenerated Clients

Commit:
Release 2023-04-10

rem7 committed:
update CF URLSigner to use crypto.Signer interface (#2087)

CloudFront URLSigner uses *rsa.PrivateKey to sign the URL. If we update
the function to receive the crypto.Signer interface (already implemented
by rsa.PrivateKey) then we can use other keys to sign. For example
hardware keys.

Commit:
Update API model

Commit:
Update endpoints model

Commit:
Regenerated Clients

Commit:
Release 2023-04-11

ztison committed:
Merge branch 'aws:main' into 2042-dynamodb-register-marshaler
